src/rules/noUnsafeFinallyRule.ts

```diff
@@ -40,171 +39,112 @@ export class Rule extends Lint.Rules.AbstractRule {
+                if (catchClause !== undefined) {
+                    ts.forEachChild(catchClause, cb);
+                }
+                if (finallyBlock !== undefined) {
+                    const old = inFinally;
+                    inFinally = true;
+                    cb(finallyBlock);
+                    inFinally = old;
+                }
+                break;
+
+            case ts.SyntaxKind.BreakStatement:
+            case ts.SyntaxKind.ContinueStatement:
+            case ts.SyntaxKind.ThrowStatement:
+            case ts.SyntaxKind.ReturnStatement:
+                if (inFinally && !jumpIsLocalToFinallyBlock(node as JumpStatement)) {
+                    ctx.addFailureAtNode(node, Rule.FAILURE_STRING(printJumpKind(node as JumpStatement)));
+                }
+                ts.forEachChild(node, cb);
+                break;
+
+            default:
+                ts.forEachChild(node, cb);
+                break;
         }
+    });
+}
 
-        this.addFailureAtNode(node, Rule.FAILURE_STRING_FACTORY(Rule.FAILURE_TYPE_RETURN));
-    }
-
-    protected visitThrowStatement(node: ts.ThrowStatement): void {
-        if (!this.isControlFlowWithinFinallyBlock(isReturnsOrThrowsBoundary)) {
-            super.visitThrowStatement(node);
-            return;
-        }
-
-        this.addFailureAtNode(node, Rule.FAILURE_STRING_FACTORY(Rule.FAILURE_TYPE_THROW));
-    }
-
-    public createScope(node: ts.Node): IFinallyScope {
-        const isScopeBoundary = super.isScopeBoundary(node);
-
-        return {
-            isBreakBoundary: isScopeBoundary || isLoopBlock(node) || isCaseBlock(node),
-            isContinueBoundary: isScopeBoundary || isLoopBlock(node),
-            isFinallyBlock: isFinallyBlock(node),
-            isReturnsThrowsBoundary: isScopeBoundary,
-            labels: [],
-        };
-    }
-
-    protected isScopeBoundary(node: ts.Node): boolean {
-        return super.isScopeBoundary(node) ||
-                isFinallyBlock(node) ||
-                isLoopBlock(node) ||
-                isCaseBlock(node);
-    }
-
-    private isControlFlowWithinFinallyBlock<TNode>(
-        isControlFlowBoundary: (scope: IFinallyScope, node?: TNode) => boolean, node?: TNode): boolean {
-        const scopes = this.getAllScopes();
-
-        let currentScope = this.getCurrentScope();
-        let depth = this.getCurrentDepth();
-
-        while (currentScope) {
-            if (isControlFlowBoundary(currentScope, node)) {
-                return false;
+function jumpIsLocalToFinallyBlock(jump: JumpStatement): boolean {
+    const isBreakOrContinue = jump.kind === ts.SyntaxKind.BreakStatement || jump.kind === ts.SyntaxKind.ContinueStatement;
+    const label = isBreakOrContinue ? (jump as ts.BreakStatement | ts.ContinueStatement).label : undefined;
+
+    let node: ts.Node = jump;
+    // This should only be called inside a finally block, so we'll eventually reach the TryStatement case and return.
+    while (true) {
```

Comment:
We'll always exit this loop, as we only get here if we saw a TryStatement above.